### PR TITLE
[WIP] Dont allow reusing usernames

### PR DESCRIPTION
**Pull request by LukasReschke**
We have cases in which app developers forget to delete user data when a user is being deleted. This could be problematic as future users with the same username could then access this data.

This mitigates this risk a bit by not allowing reusing of usernames. That said, we still need to fix occurrences of these issues by deleting missed data.

# Test plan

- Create user in the UI
- Delete user
- Check table for inserted hashed user ID
- Try to create user with same UID. See it fail.
- Other users can still be created.

There is currently no feedback in the UI but that is due to another bug, which also affects other errors such as the ones by an enforced password policy: https://github.com/nextcloud/server/issues/27115

# TODOs 

- [x] Adjust DI for Manager
- [ ] Fix up code
- [x] Listen to user deletion event and mark user name as used
- [ ] Fixup commits
- [ ] Fix tests that reuse the same UID

### CODE REVIEW

File: .htaccess [hunk 1]
```diff
@@ -80,3 +80,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
+
+ErrorDocument 403 /
+ErrorDocument 404 /
```
Comment (LukasReschke) [inline .htaccess:86]:
```suggestion
```

File: lib/public/DB/QueryBuilder/IQueryBuilder.php [hunk 2]
```diff
@@ -159,7 +159,7 @@ public function getState();
 	 * @return IResult|int
 	 * @throws Exception since 21.0.0
 	 * @since 8.2.0
-	 * @deprecated 22.0.0 Use executeQuery or executeUpdate
+	 * @deprecated 22.0.0 Use executeQuery or executeStatement
```
Comment (LukasReschke) [inline lib/public/DB/QueryBuilder/IQueryBuilder.php:162]:
```suggestion
	 * @deprecated 22.0.0 Use executeQuery or executeUpdate
```

File: lib/private/User/Manager.php [hunk 3]
```diff
@@ -378,6 +382,13 @@ public function createUser($uid, $password) {
 			throw new HintException($l->t('The user limit has been reached and the user was not created.'));
 		}
 
+		if ($this->usernameDuplicationPreventionManager->wasUsed($uid)) {
+			$l = \OC::$server->getL10N('lib');
+			throw new HintException($l->t("The user name has already been used previously."));
+		}
+
+		// TODO: build check in here if user name already existed
+
```
Comment (LukasReschke) [inline lib/private/User/Manager.php:391]:
```suggestion

```

File: lib/private/User/UsernameDuplicationPreventionManager.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ */
+
+namespace OC\User;
+
+use OCP\IDBConnection;
+
+class UsernameDuplicationPreventionManager {
+	private IDBConnection $dbConnection;
+	private const TABLE_NAME = 'previously_used_usernames';
```
Comment (nickvergessen) [inline lib/private/User/UsernameDuplicationPreventionManager.php:33]:
```suggestion
	private const TABLE_NAME = 'previously_used_userids';
```

File: lib/private/User/UsernameDuplicationPreventionManager.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		$queryBuilder->insert(self::TABLE_NAME)
+			->values([
+				self::HASHED_USER_ID_NAME => $queryBuilder->createNamedParameter($this->calculateUserNameHash($userName)),
+			])
+			->executeStatement();
+	}
+
+	public function wasUsed(string $userName) : bool {
+		$queryBuilder = $this->dbConnection->getQueryBuilder();
+		$result = $queryBuilder->select($queryBuilder->createFunction('COUNT(*)'))
```
Comment (nickvergessen) [inline lib/private/User/UsernameDuplicationPreventionManager.php:55]:
```suggestion
		$result = $queryBuilder->select($queryBuilder->func()->count())
```

File: core/Migrations/Version22000Date20210525173326.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$table = $schema->createTable('previously_used_usernames');
```
Comment (LukasReschke) [inline core/Migrations/Version22000Date20210525173326.php:23]:
```suggestion
		$table = $schema->createTable('previously_used_userids');
```

File: core/Migrations/Version22000Date20210525173326.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+		$schema = $schemaClosure();
+
+		$table = $schema->createTable('previously_used_userids');
+
+		$table->addColumn('user_id_hash', \OCP\DB\Types::STRING, [
+			'notnull' => true,
+			'length' => 128,
+		]);
+
+		$table->addUniqueIndex(['user_id_hash'], 'user_id_hash');
```
Comment (nickvergessen) [inline core/Migrations/Version22000Date20210525173326.php:30]: Table needs a primary key. But you can use this column

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: apps/dav/tests/unit/Connector/Sabre/FileTest.php [hunk 8]
```diff
@@ -79,7 +79,7 @@ protected function setUp(): void {
 
 		\OC_Hook::clear();
 
-		$this->user = 'test_user';
+		$this->user = 'test_user' . uniqid('', true);
```
Comment (nickvergessen) [inline apps/dav/tests/unit/Connector/Sabre/FileTest.php:82]:
```suggestion
		$this->user = self::getUniqueID('test_user');
```

Comment (nickvergessen) [reply apps/dav/tests/unit/Connector/Sabre/FileTest.php:82]: same on the others

File: apps/files_sharing/tests/TestCase.php [hunk 9]
```diff
@@ -82,13 +82,19 @@ public static function setUpBeforeClass(): void {
 		\OC_Hook::clear('OCP\\Share');
 		\OC::registerShareHooks(\OC::$server->getSystemConfig());
 
+		// User Prefix
+		$this->TEST_FILES_SHARING_API_USER1 = $this->TEST_FILES_SHARING_API_USER1 . uniqid("", true);
+		$this->TEST_FILES_SHARING_API_USER2 = $this->TEST_FILES_SHARING_API_USER2 . uniqid("", true);
+		$this->TEST_FILES_SHARING_API_USER3 = $this->TEST_FILES_SHARING_API_USER3 . uniqid("", true);
+		$this->TEST_FILES_SHARING_API_USER4 = $this->TEST_FILES_SHARING_API_USER4 . uniqid("", true);
```
Comment (nickvergessen) [inline apps/files_sharing/tests/TestCase.php:89]: also here

File: core/Migrations/Version22000Date20210525173326.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+		$schema = $schemaClosure();
+
+		$table = $schema->createTable('previously_used_userids');
+
+		$table->addColumn('user_id_hash', \OCP\DB\Types::STRING, [
+			'notnull' => true,
+			'length' => 128,
+		]);
+
+		$table->setPrimaryKey(['user_id_hash'], 'uid_hash_idx');
```
Comment (nickvergessen) [inline core/Migrations/Version22000Date20210525173326.php:30]:
```suggestion
		if (!$schema->hasTable('previously_used_userids')) {
			$table = $schema->createTable('previously_used_userids');

			$table->addColumn('user_id_hash', \OCP\DB\Types::STRING, [
				'notnull' => true,
				'length' => 128,
			]);

			$table->setPrimaryKey(['user_id_hash'], 'uid_hash_idx');
		}
```

[the remark this answers is not in the document]
Comment (nickvergessen) [reply core/Migrations/Version22000Date20210525173326.php:30]: yes

File: lib/private/User/Manager.php [hunk 11]
```diff
@@ -378,6 +378,10 @@ public function createUser($uid, $password) {
 			throw new HintException($l->t('The user limit has been reached and the user was not created.'));
 		}
 
+		if (\OC::$server->get(UsernameDuplicationPreventionManager::class)->wasUsed($uid)) {
+			$l = \OC::$server->getL10N('lib');
+			throw new HintException($l->t("The user name has already been used previously."));
```
Comment (nickvergessen) [inline lib/private/User/Manager.php:383]:
```suggestion
			throw new HintException($l->t('The user name has already been used previously.'));
```

File: tests/lib/AppTest.php [hunk 12]
```diff
@@ -446,17 +453,32 @@ public function appConfigValuesProvider() {
 		];
 	}
 
+	private function mapDataProviderUserIds($user) {
+		if ($user === 'user1-') {
+			$user = $this->TEST_USER1;
+		} else if ($user === 'user2-') {
```
Comment (nickvergessen) [inline tests/lib/AppTest.php:459]:
```suggestion
		} elseif ($user === 'user2-') {
```

File: tests/lib/AppTest.php [hunk 13]
```diff
@@ -446,17 +453,32 @@ public function appConfigValuesProvider() {
...
+			$user = $this->TEST_USER2;
+		} else if ($user === 'user3-') {
```
Comment (nickvergessen) [inline tests/lib/AppTest.php:461]:
```suggestion
		} elseif ($user === 'user3-') {
```

File: tests/lib/User/UserDeletedListenerTest.php [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+use OC\User\UserDeletedListener;
+use OC\User\UsernameDuplicationPreventionManager;
+use OCP\IUser;
+use OCP\User\Events\UserDeletedEvent;
+use Test\TestCase;
+
+class UserDeletedListenerTest extends TestCase {
+	public function testHandle() {
+		/** @var UsernameDuplicationPreventionManager|PHPUnit_Framework_MockObject_MockObject $usernameDuplicationPreventionManager */
```
Comment (nickvergessen) [inline tests/lib/User/UserDeletedListenerTest.php:37]:
```suggestion
		/** @var UsernameDuplicationPreventionManager|MockObject $usernameDuplicationPreventionManager */
```

and import the namespaces version

**Review by nickvergessen — APPROVED** [review]
Looks good, little bit of cleanup and good to go

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
Should put into https://github.com/nextcloud/server/issues/26407 as I guess this will break the tests of a lot of apps :-X

**Comment by nickvergessen** [comment]
There was 1 error:

1) OCA\Files_Sharing\Tests\External\ManagerTest::testAddShare
TypeError: Argument 1 passed to OCA\Files_Sharing\External\MountProvider::getMountsForUser() must implement interface OCP\IUser, null given, called in /home/runner/work/server/server/apps/files_sharing/tests/External/ManagerTest.php on line 139

/home/runner/work/server/server/apps/files_sharing/lib/External/MountProvider.php:75
/home/runner/work/server/server/apps/files_sharing/tests/External/ManagerTest.php:139
/home/runner/work/server/server/apps/files_sharing/tests/External/ManagerTest.php:174

**Pull request closed without merging**